File: src/main/java/zechs/zplex/sync/data/remote/interceptors/TmdbApiKeyInterceptor.java
package zechs.zplex.sync.data.remote.interceptors;

import okhttp3.HttpUrl;
import okhttp3.Interceptor;
import okhttp3.Request;
import okhttp3.Response;
import org.springframework.stereotype.Component;

import java.io.IOException;

import static zechs.zplex.sync.utils.Constants.TMDB_API_KEY;
import static zechs.zplex.sync.utils.Constants.TMDB_API_URL;

@Component
public final class TmdbApiKeyInterceptor implements Interceptor {

    private static final String TMDB_API_KEY_QUERY_PARAM = "api_key";

    @Override
    public Response intercept(Chain chain) throws IOException {
        Request request = chain.request();
        String requestUrl = request.url().toString();

        if (requestUrl.startsWith(TMDB_API_URL)) {
            HttpUrl newRequestUrl = addApiKeyIfNotPresent(request);
            Request newRequest = request.newBuilder().url(newRequestUrl).build();
            return chain.proceed(newRequest);
        }
        return chain.proceed(request);
    }

    private HttpUrl addApiKeyIfNotPresent(Request request) {
        HttpUrl requestUrl = request.url();
        boolean hasApiKey = requestUrl.queryParameter(TMDB_API_KEY_QUERY_PARAM) != null;
        if (!hasApiKey) {
            return requestUrl.newBuilder().addQueryParameter(TMDB_API_KEY_QUERY_PARAM, TMDB_API_KEY).build();
        }
        return requestUrl;
    }
}

File: src/main/kotlin/utils/TmdbApiKeyInterceptor.kt
package zechs.zplex.sync.utils

import okhttp3.HttpUrl
import okhttp3.Interceptor
import okhttp3.Request
import okhttp3.Response
import zechs.zplex.sync.utils.Constants.Companion.TMDB_API_URL

class TmdbApiKeyInterceptor(
    private val apiKey: String
) : Interceptor {

    override fun intercept(chain: Interceptor.Chain): Response {
        val request = chain.request()
        val url = request.url.toString()

        return if (url.startsWith(TMDB_API_URL)) {
            val newRequestUrl = addApiKeyIfNotPresent(request)
            val newRequest = request.newBuilder().url(newRequestUrl).build()
            return chain.proceed(newRequest)
        } else {
            chain.proceed(request)
        }
    }

    private fun addApiKeyIfNotPresent(request: Request): HttpUrl {
        val url = request.url
        val hasApiKey = url.queryParameter("api_key") != null
        return if (hasApiKey) {
            url
        } else {
            url.newBuilder().addQueryParameter("api_key", apiKey).build()
        }
    }

}